File: frontend/src/mocks/hanzi.ts
// Mock 数据：Step 4 前端实现阶段的假数据，联调后可删除
import type { HanziCharacter, HanziLevel } from '../types/hanzi'

const NOW = '2026-07-01T00:00:00Z'

export const mockLevels: HanziLevel[] = [
  {
    id: 'lvl-1',
    name: '启蒙 · Level 1',
    description: '最基础的常用字：人、口、日、月、山、水……',
    order_index: 1,
    total: 20,
    learned: 8,
    created_at: NOW,
    updated_at: NOW,
  },
  {
    id: 'lvl-2',
    name: '启蒙 · Level 2',
    description: '巩固扩展：数字、方位、颜色',
    order_index: 2,
    total: 40,
    learned: 0,
    created_at: NOW,
    updated_at: NOW,
  },
  {
    id: 'lvl-3',
    name: '进阶 · Level 3',
    description: '日常场景：家庭、食物、动物',
    order_index: 3,
    total: 80,
    learned: 0,
    created_at: NOW,
    updated_at: NOW,
  },
]

const level1Chars = [
  '人', '口', '日', '月', '山', '水', '火', '土', '木', '金',
  '大', '小', '上', '下', '中', '天', '子', '女', '手', '足',
]

export const mockCharacters: Record<string, HanziCharacter[]> = {
  'lvl-1': level1Chars.map((char, i) => ({
    id: `char-1-${i}`,
    level_id: 'lvl-1',
    char,
    pinyin: [
      'rén', 'kǒu', 'rì', 'yuè', 'shān', 'shuǐ', 'huǒ', 'tǔ', 'mù', 'jīn',
      'dà', 'xiǎo', 'shàng', 'xià', 'zhōng', 'tiān', 'zǐ', 'nǚ', 'shǒu', 'zú',
    ][i],
    example_words: [
      ['人类', '大人', '人口', '人们'],
      ['口水', '开口', '门口', '人口'],
      ['日子', '生日', '日出', '今日'],
      ['月亮', '月光', '每月', '岁月'],
      ['山水', '山上', '高山', '山口'],
      ['水果', '喝水', '山水', '水口'],
      ['火车', '大火', '着火', '火山'],
      ['土地', '泥土', '土豆', '尘土'],
      ['木头', '树木', '木门', '木料'],
      ['金子', '金色', '五金', '金鱼'],
      ['大小', '大人', '大门', '大家'],
      ['小心', '小孩', '小时', '大小'],
      ['上面', '上班', '早上', '向上'],
      ['下面', '下雨', '下班', '下午'],
      ['中间', '中国', '中心', '空中'],
      ['天空', '今天', '天气', '每天'],
      ['孩子', '儿子', '子女', '种子'],
      ['女孩', '女人', '子女', '少女'],
      ['手指', '双手', '拍手', '手心'],
      ['脚步', '双脚', '脚下', '脚印'],
    ][i],
    order_index: i,
    learned: i < 8,
    learned_at: i < 8 ? NOW : null,
    created_at: NOW,
    updated_at: NOW,
  })),
  'lvl-2': [],
  'lvl-3': [],
}

File: backend/app/application/dto/hanzi.py
"""汉字模块出参 DTO。

user 侧与 admin 侧字段**故意不同**，与改造前的手写 dict 保持一致：
  - 级别：user 侧多 `learned`（该用户已学字数），admin 侧只有 `total`；
  - 字条：user 侧带 `learned` / `learned_at`，admin 侧带 `created_at` / `updated_at`。
不要图省事把两边合并成一个模型——那会给对方端点凭空加字段。
"""

from __future__ import annotations

from app.application.dto.base import DTO, IsoDtOpt
from app.domain.models.hanzi import HanziCharacter, HanziLevel, HanziUserProgress


class LevelBrief(DTO):
    """字表接口里的级别摘要（user / admin 两侧同形，无时间字段）。"""

    id: str
    name: str
    description: str | None
    order_index: int


class LevelForUser(DTO):
    id: str
    name: str
    description: str | None
    order_index: int
    total: int
    learned: int
    created_at: IsoDtOpt
    updated_at: IsoDtOpt

    @classmethod
    def build(cls, level: HanziLevel, total: int, learned: int) -> LevelForUser:
        return cls(
            id=level.id,
            name=level.name,
            description=level.description,
            order_index=level.order_index,
            total=total,
            learned=learned,
            created_at=level.created_at,
            updated_at=level.updated_at,
        )


class LevelForAdmin(DTO):
    id: str
    name: str
    description: str | None
    order_index: int
    total: int
    created_at: IsoDtOpt
    updated_at: IsoDtOpt

    @classmethod
    def build(cls, level: HanziLevel, total: int) -> LevelForAdmin:
        return cls(
            id=level.id,
            name=level.name,
            description=level.description,
            order_index=level.order_index,
            total=total,
            created_at=level.created_at,
            updated_at=level.updated_at,
        )


class CharacterForUser(DTO):
    id: str
    char: str
    pinyin: str
    example_words: list[str]
    order_index: int
    learned: bool
    learned_at: IsoDtOpt

    @classmethod
    def build(
        cls, character: HanziCharacter, progress: HanziUserProgress | None
    ) -> CharacterForUser:
        return cls(
            id=character.id,
            char=character.char,
            pinyin=character.pinyin,
            example_words=character.example_words or [],
            order_index=character.order_index,
            learned=progress is not None,
            learned_at=progress.learned_at if progress else None,
        )


class CharacterForAdmin(DTO):
    id: str
    char: str
    pinyin: str
    example_words: list[str]
    order_index: int
    created_at: IsoDtOpt
    updated_at: IsoDtOpt

    @classmethod
    def build(cls, character: HanziCharacter) -> CharacterForAdmin:
        return cls(
            id=character.id,
            char=character.char,
            pinyin=character.pinyin,
            example_words=character.example_words or [],
            order_index=character.order_index,
            created_at=character.created_at,
            updated_at=character.updated_at,
        )


class ProgressOut(DTO):
    character_id: str
    learned: bool
    learned_at: IsoDtOpt
